aws/s3.go

```diff
@@ -0,0 +1,295 @@
+		Bucket: bucketName,
+	}
+
+	tags := []map[string]string{}
+
+	result, err := svc.GetBucketTagging(input)
+	if err != nil {
+		return tags, err
+	}
+
+	for _, tagSet := range result.TagSet {
+		tags = append(tags, map[string]string{"Key": *tagSet.Key, "Value": *tagSet.Value})
+	}
+
+	return tags, err
+}
+
+// getAllS3Buckets lists and returns a map of per region AWS S3 buckets which were created before excludeAfter
+func getAllS3Buckets(session *session.Session, excludeAfter time.Time, bucketNameSubStr string) (map[string][]*string, error) {
+	svc := s3.New(session)
+
+	input := &s3.ListBucketsInput{}
+
+	output, err := svc.ListBuckets(input)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+
+OUTER:
+	for _, bucket := range output.Buckets {
+		logging.Logger.Debugf("Checking - Bucket %s", *bucket.Name)
+
+		if len(bucketNameSubStr) > 0 {
+			if !strings.Contains(*bucket.Name, bucketNameSubStr) {
+				logging.Logger.Debugf("Skipping - Bucket %s - failed substring filter - %s", *bucket.Name, bucketNameSubStr)
+				continue
+			}
+		}
+
+		if !excludeAfter.After(*bucket.CreationDate) {
+			logging.Logger.Debugf("Skipping - Bucket %s - matched CreationDate filter", *bucket.Name)
+			continue
+		}
+
+		bucketTags, err := getS3BucketTags(svc, bucket.Name)
+		if len(bucketTags) > 0 {
+			for _, tagSet := range bucketTags {
+				if tagSet["Key"] == "cloud-nuke-excluded" && tagSet["Value"] == "true" {
+					logging.Logger.Infof("Skipping - Bucket %s - matched tag filter", *bucket.Name)
+					continue OUTER
+				}
+			}
+		}
+
+		bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
+		if err != nil {
+			logging.Logger.Warnf("Skipping - Bucket %s - Failed to get bucket location.", *bucket.Name)
+			continue
+		}
+
+		if _, ok := bucketNamesPerRegion[bucketRegion]; !ok {
+			bucketNamesPerRegion[bucketRegion] = []*string{}
+		}
+
+		bucketNamesPerRegion[bucketRegion] = append(bucketNamesPerRegion[bucketRegion], bucket.Name)
+	}
+
+	return bucketNamesPerRegion, nil
+}
+
+func getS3BucketObjectVersions(svc *s3.S3, bucketName *string, objectKey *string) ([]*string, error) {
+	versions := []*string{}
+
+	err := svc.ListObjectVersionsPages(
+		&s3.ListObjectVersionsInput{
+			Bucket: bucketName,
+			Prefix: objectKey,
+		},
+		func(page *s3.ListObjectVersionsOutput, lastPage bool) (shouldContinue bool) {
+			for _, obj := range page.Versions {
+				versions = append(versions, obj.VersionId)
+			}
+			return true
+		},
+	)
+	return versions, err
+}
+
+func getS3BucketObjects(svc *s3.S3, bucketName *string, isVersioned bool) ([]*s3.ObjectIdentifier, error) {
+	identifiers := []*s3.ObjectIdentifier{}
+	hasError := false
+
+	err := svc.ListObjectsV2Pages(
+		&s3.ListObjectsV2Input{
+			Bucket:  bucketName,
+			MaxKeys: aws.Int64(3),
+		},
+
+		func(page *s3.ListObjectsV2Output, lastPage bool) (shouldContinue bool) {
+
+			for _, obj := range page.Contents {
+				if isVersioned {
```

Comment:
If versioned, we should go directly to the `ListObjectVersionsPages` instead of iterating on the objects, since `ListObjectVersionsPages` without a `Prefix` will list everything in the bucket. This not only simplifies the code here, but also reduces the API calls down to one vs N, where N is the number of objects.

Reply:
Done. Thanks for catching that.